### PR TITLE
Distributed randomized SVD

### PR DESCRIPTION
**Pull request by mrfh92**
## Due Diligence

- General:
    - [x]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [x] unit tests: all split configurations tested
    - [x] unit tests: multiple dtypes tested
    - [x] documentation updated where needed

## Description

Adds randomized SVD and corresponding interface in PCA 

Issue/s resolved: #1457 

## Type of change

- New feature (non-breaking change which adds functionality)

#### Does this change modify the behaviour of other functions? If so, which?
yes (slightly, of PCA) / no (all other functionality)

### CODE REVIEW

File: heat/core/random.py [hunk 1]
```diff
@@ -129,8 +129,8 @@ def __counter_sequence(
         c_0 = (__counter & (max_count << 64)) >> 64
     c_1 = __counter & max_count
     total_elements = torch.prod(torch.tensor(shape))
-    if total_elements.item() > 2 * max_count:
-        raise ValueError(f"Shape is to big with {total_elements} elements")
+    # if total_elements.item() > 2 * max_count:
+    #    raise ValueError(f"Shape is to big with {total_elements} elements")
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/random.py:133]: actually this needs to be kept, but without being commented out

File: heat/decomposition/pca.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def __init__(
         whiten: bool = False,
         svd_solver: str = "hierarchical",
         tol: Optional[float] = None,
-        iterated_power: Union[str, int] = "auto",
+        iterated_power: Union[str, int] = 0,
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/decomposition/pca.py:88]: done

[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/decomposition/pca.py:88]: docs for PCA are now complete

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Looks great, only a couple tiny changes needed as far as I'm concerned. Thanks @mrfh92 !

File: heat/core/random.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/random.py]: has been resolved meanwhile

File: heat/core/linalg/svdtools.py [hunk 3]
```diff
@@ -21,7 +21,21 @@
...
-__all__ = ["hsvd_rank", "hsvd_rtol", "hsvd"]
+__all__ = ["hsvd_rank", "hsvd_rtol", "hsvd", "rsvd"]
+
+
+def _check_SVD_input(A):
+    if not isinstance(A, DNDarray):
+        raise TypeError(f"Argument needs to be a DNDarray but is {type(A)}.")
+    if not A.ndim == 2:
+        raise ValueError("A needs to be a 2D matrix")
+    if not A.dtype == types.float32 and not A.dtype == types.float64:
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/core/linalg/svdtools.py]: I have introduced `heat_type_is_realfloating` (as there already existed `heat_type_is_complexfloating`)

[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/linalg/svdtools.py]: @mrfh92 did you push the modified `types` module?

File: heat/core/linalg/svdtools.py [hunk 4]
```diff
@@ -529,3 +525,92 @@ def compute_local_truncated_svd(
...
+        2D-array (float32/64) of which the rSVD has to be computed.
+    rank : int
+        truncation rank. (This parameter corresponds to `n_components` in scikit-learn's TruncatedSVD.)
+    n_oversamples : int, optional
+        number of oversamples. The default is 10.
+    power_iter : int, optional
+        number of power iterations. The default is 0.
+        Choosing `power_iter > 0` can improve the accuracy of the SVD approximation in the case of slowly decaying singular values, but increases the computational cost.
+    qr_procs_to_merge : int, optional
+        number of processes to merge at each step of QR decomposition in the power iteration (if power_iter > 0). The default is 2. See the corresponding remarks for `heat.linalg.qr` for more details.
```
Comment (ClaudiaComito) [inline heat/core/linalg/svdtools.py:560]:
```suggestion
        number of processes to merge at each step of QR decomposition in the power iteration (if power_iter > 0). The default is 2. See the corresponding remarks for :func:`heat.linalg.qr() <heat.core.linalg.qr.qr()>` for more details.
```

Not 100% sure but I think this is the way to render links to other functions on ReadTheDocs.

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks for addressing all the comments, I found a few typos, otherwise this looks great.

**Comment by mrfh92** [comment]
@ClaudiaComito thanks for reading so carefully that you found all the typos I had overlooked :+1:

**Review by ClaudiaComito — APPROVED** [review]
Great, thanks @mrfh92 !

**Review by ClaudiaComito — APPROVED** [review]

**Pull request merged**